### DISCUSSION TITLE
Capturing PowerShell console output

### DISCUSSION THREAD

**Discussion by stolaf** [TomasBouda/PoShLog#25] [General]
it is possible to log the whole console output without Write-*Log so that unhandled exceptions are recorded?

**Comment by TomasBouda** [comment]
Hi, I'm not sure if I understand that correctly. Can you please show me example of what you are trying to accomplish? thx

**Edit**
You could use [trap](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_trap?view=powershell-7.2) keyword to catch terminating error. If you want to hide powershell error messages completely, you would have to change [`$ErrorActionPreference` variable](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.2#erroractionpreference) to `SilentlyContinue`.

```ps1
Import-module PoShLog

New-Logger |
Add-SinkConsole | 
Start-Logger

# Change EA preference variable to SilentlyContinue
$previsousEAP = $ErrorActionPreference
$ErrorActionPreference = 'SilentlyContinue'

# Trap setup to log unhandled exceptions using PoShLog
trap {
	Write-ErrorLog -MessageTemplate $_ -ErrorRecord $_
}

# Your script starts here

# Some code that throws an error
1/$null

# Your script ends here

# Change EA preference variable back to it's previous state
$ErrorActionPreference = $previsousEAP

Close-Logger
```
Previous code would produce similar output:

```
[10:55:23 ERR] Attempted to divide by zero.
System.DivideByZeroException: Attempted to divide by zero.
```

**Comment by stolaf** [comment]
Hi Tomas,
your example works fine, however this not work:

New-Logger |
Add-SinkConsole | 
Add-SinkFile -Path 'C:\Temp\MyAwesome.log' | 
Start-Logger

$previsousEAP = $ErrorActionPreference
$ErrorActionPreference = 'SilentlyContinue'

trap {
    Write-ErrorLog -MessageTemplate $_ -ErrorRecord $_
}

Get-ChildItem -Path 'C:\Path_not_exist' -Filter *.txt
1 / $null

$ErrorActionPreference = $previsousEAP

Close-Logger

**Reply by TomasBouda** [reply]
I think your problem is that `trap` only catches **terminating errors**. You would have to add `-ErrorAction Stop` to `Get-ChildItem` cmdlet or execute `$ErrorActionPreference = 'Stop'`:

```ps1
New-Logger |
	Add-SinkConsole |
	Add-SinkFile -Path 'C:\Temp\MyAwesome.log' |
	Start-Logger

$previsousEAP = $ErrorActionPreference
$ErrorActionPreference = 'SilentlyContinue'

trap {
	Write-ErrorLog -MessageTemplate $_ -ErrorRecord $_
}

Get-ChildItem -Path 'C:\Path_not_exist' -Filter *.txt -ErrorAction Stop
1 / $null

$ErrorActionPreference = $previsousEAP

Close-Logger
```

Wich produces:

```
[16:42:02 ERR] Cannot find path 'C:\Path_not_exist' because it does not exist.
System.Management.Automation.ItemNotFoundException: Cannot find path 'C:\Path_not_exist' because it does not exist.
   at System.Management.Automation.SessionStateInternal.GetChildItems(String path, Boolean recurse, UInt32 depth, CmdletProviderContext context)
   at System.Management.Automation.ChildItemCmdletProviderIntrinsics.Get(String path, Boolean recurse, UInt32 depth, CmdletProviderContext context)
   at Microsoft.PowerShell.Commands.GetChildItemCommand.ProcessRecord()
[16:42:02 ERR] Attempted to divide by zero.
System.DivideByZeroException: Attempted to divide by zero.
```

Please next time be more specific, "this doesn't work" usualy doesn't give me enough information. Best is, to report your PowerShell version, script output and expected behaviour. Thanks 🙏

**Comment by stolaf** [comment]
Hi Tomas,

Is there another way without $ErrorActionPreference, $WarningPreference etc. to write the console output to a log? I often use foreign scripts and modules and don't want to make any customizations there ...

**Reply by TomasBouda** [reply]
Hi again 😄 

I did some thinking and experimentations and I've come up with a solution how to redirect messages from third-party script/cmdlet that is being called. I'm not sure if it suits your needs, but it could at least help you a little to find the perfect solution. Keep in mind that it's experimental code and I didn't test it enough, so I would be carefull running it in production environment. 

Consider following "third-party" script(*lets call it `test.ps1`*) we want to execute and redirect it's output:

`.\test.ps1`
```ps1
[Cmdletbinding()]
param(
	[bool]$WriteMessages
)

function Write-Messages {
	Write-Verbose 'Verbose message'
	Write-Debug 'Debug'
	Write-Information 'Information message'
	Write-Host 'host only'
	Write-Warning 'Some warning'
	Write-Error 'Error message'
	Write-Output 'Some output'
}

if($WriteMessages){
	Write-Messages
}
```

Now let's create our main script that will catch messages from previous script and log them using PoShLog(*both scripts are in same directory*):

`.\main.ps1`
```ps1
New-Logger |
	Set-MinimumLevel -Value Verbose -ToPreference | # -ToPreference is importat in order to redirect all streams
	Add-SinkConsole |
	Start-Logger

# trap from previous example is not essential in this case
trap {
	Write-ErrorLog -MessageTemplate $_ -ErrorRecord $_
}

function Out-Log {
	[Cmdletbinding()]
	param(
		[Parameter(Mandatory = $true, ValueFromPipeline = $true)]
		$InputObject
	)

	process {
		if ($InputObject.GetType().Name -eq 'VerboseRecord') {
			Write-VerboseLog $InputObject
		}
		elseif ($InputObject.GetType().Name -eq 'DebugRecord') {
			Write-DebugLog $InputObject
		}
		elseif ($InputObject.GetType().Name -eq 'InformationRecord') {
			Write-InfoLog $InputObject
		}
		elseif ($InputObject.GetType().Name -eq 'WarningRecord') {
			Write-WarningLog $InputObject
		}
		elseif ($InputObject.GetType().Name -eq 'ErrorRecord') {
			Write-ErrorLog $InputObject -ErrorRecord $InputObject
		}
		else {
			$InputObject
		}
	}
}

# Dot source example script
. "$PSScriptRoot\test.ps1"

# Example 1
# Call third party cmdlet and redirect it's output to PoShLog
$cmdletOutput = Write-Messages *>&1 | Out-Log

# Example 2
# Call third party script directly and redirect it's output to PoShLog
$scriptOutput = & "$PSScriptRoot\test.ps1" $true *>&1 | Out-Log

Close-Logger
```

When we run script `.\main.ps1` we should get similar output:

```
[11:46:39 VRB] Verbose message
[11:46:39 DBG] Debug
[11:46:39 INF] Information message
[11:46:39 INF] host only
[11:46:39 WRN] Some warning
[11:46:39 ERR] Error message
Microsoft.PowerShell.Commands.WriteErrorException: Error message
[11:46:39 VRB] Verbose message
[11:46:39 DBG] Debug
[11:46:39 INF] Information message
[11:46:39 INF] host only
[11:46:39 WRN] Some warning
[11:46:39 ERR] Error message
Microsoft.PowerShell.Commands.WriteErrorException: Error message
```

Messages above are `Write-*` messages from script `.\test.ps1`, these messages get [redirected](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_redirection?view=powershell-7.2) using `*>&1` operator and they are processed/filtered in `Out-Log` cmdlet which passes the messages to correct PoShLog's `Write-*Log` cmdlet. If you want to log also debug and verbose messages, like in example above, **don't forget to change PoShLog's minimum level** using `Set-MinimumLevel -Value Verbose -ToPreference` and **most importantly pass this value to [preference variables](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.2#erroractionpreference) using `-ToPreference`**.

I hope this helps. If you foud any bugs or aspects I forget think about, let me know. Good luck ☘️

**Comment by MrHinsh** [comment]
@TomasBouda how would I capture other outputs?

For xamle:

```
Write-InfoLog "Run Compile & Test"

dotnet restore
dotnet build /p:Version=$($versionInfo.AssemblySemVer) /p:FileVersion=$($versionInfo.AssemblySemVer) /p:InformationalVersion=$($versionInfo.InformationalVersion)
```

Results in everything output from `dotnet restore` being lost from the log

**Reply by TomasBouda** [reply]
Hi @MrHinsh,

I know this is not elegant solution, but what I tried and what works is doing this:
```ps1
Write-InfoLog "Run Compile & Test"

dotnet --list-runtimes | Write-InfoLog

# or

Invoke-Command -ScriptBlock {
	dotnet --list-runtimes
	dotnet --list-sdks
	# Or whatever you want to do
} | Write-InfoLog
```

Let me know if that's solution for you. In the meantime I'll think about some "global" redirect I'm thinking about ressurecting this project anyways 😄

**Reply by MrHinsh** [reply]
![image](https://github.com/PoShLog/PoShLog/assets/5205575/5d327e8f-9e7b-4042-a798-ee179dbf5e4b)
This is all the log I get for:
```
Write-InfoLog "Run Compile & Test"

dotnet restore | Write-InfoLog
dotnet build /p:Version=$($versionInfo.AssemblySemVer) /p:FileVersion=$($versionInfo.AssemblySemVer) /p:InformationalVersion=$($versionInfo.InformationalVersion) | Write-InfoLog

dotnet test --collect "Code coverage" --no-build --filter "(TestCategory=L0|TestCategory=L1)" --logger trx --results-directory .\output\testresults | Write-InfoLog
dotnet test --collect "Code coverage" --no-build --filter "(TestCategory=L2|TestCategory=L3)" --logger trx --results-directory .\output\testresults | Write-InfoLog
dotnet test --collect "Code coverage" --no-build --filter "(TestCategory!=L0&TestCategory!=L1&TestCategory!=L2&TestCategory!=L3)" --logger trx --results-directory .\output\testresults | Write-InfoLog
```

There shojld be a lot more.

**Reply by TomasBouda** [reply]
Ok, let's isolate your example to `dotnet build`
When you run 
```ps1
dotnet build /p:Version=$($versionInfo.AssemblySemVer) /p:FileVersion=$($versionInfo.AssemblySemVer) /p:InformationalVersion=$($versionInfo.InformationalVersion)
```
I guess there is some output in console...

when you run
```ps1
New-Logger | Add-SinkConsole | Start-Logger

dotnet build /p:Version=$($versionInfo.AssemblySemVer) /p:FileVersion=$($versionInfo.AssemblySemVer) /p:InformationalVersion=$($versionInfo.InformationalVersion) | Write-InfoLog

Close-Logger
```
Do you get same output as in previous case, or any output whatsoever?

You could also try redirecting all streams using `*>&1`, but I think it not gonna help in this case.
```ps1
...
dotnet build /p:Version=$($versionInfo.AssemblySemVer) /p:FileVersion=$($versionInfo.AssemblySemVer) /p:InformationalVersion=$($versionInfo.InformationalVersion) *>&1 | Write-InfoLog
...
```

Also can you please run this snippet:
```
$PSVersionTable>>info.txt;$Host>>info.txt;[System.Environment]::OSVersion>>info.txt;
```
and send me content of `.\info.txt`, thx.
Are you running it inside azure devops agent perhaps?
